### ISSUE TITLE
Handle JSONRPC disconnections gracefully

### ISSUE THREAD

**Issue by romanz** [romanz/electrs#18]
Currently `electrs` panics on any disconnection from `bitcoind`.

**Comment by greenaddress** [comment]
more instances

```
2018-07-16T00:52:28.407+00:00 - ERROR - server failed: Error: RPC failed: []                                                                                                                                       
Caused by: request failed: {"result":null,"error":{"code":-5,"message":"Transaction not in mempool"},"id":null}                                                                                                    
stack backtrace:                                                                                                                                                                                                   
   0:     0x55f75180ddac - backtrace::backtrace::trace::h44f27ad6608f02da                                                                                                                                          
   1:     0x55f75180d402 - backtrace::capture::Backtrace::new_unresolved::hb555da2dec445946                                                                                                                        
   2:     0x55f75180ccd7 - error_chain::backtrace::imp::InternalBacktrace::new::hc1c04021afee31b3                                                                                                                  
   3:     0x55f75180d044 - <error_chain::State as core::default::Default>::default::he24c386d1d298d64                                                                                                              
   4:     0x55f751748cf8 - electrs::daemon::Daemon::call_jsonrpc::h5148baa35ebacdad                                                                                                                                
   5:     0x55f75174978a - electrs::daemon::Daemon::requests::h9854bd24136c8b73                                                                                                                                    
   6:     0x55f75174beed - electrs::daemon::Daemon::gettransactions::h93b25cda33204a41                                                                                                                             
   7:     0x55f75178ace0 - electrs::mempool::Tracker::update::h9fe07ffc4b13e9f1                                                                                                                                    
   8:     0x55f75176802b - electrs::query::Query::update_mempool::h5a3567190cf4c2bb                                                                                                                                
   9:     0x55f751739df6 - electrs::main::h5294e30fc465d8cf                                                                                                                                                        
  10:     0x55f751738782 - std::rt::lang_start::{{closure}}::hc74e7e3dcbd83f22                                                                                                                                     
  11:     0x55f751bea4b2 - std::rt::lang_start_internal::{{closure}}::h72602870b4a34363                                                                                                                            
                        at libstd/rt.rs:59                                                                                                                                                                         
                         - std::panicking::try::do_call::hc2917ab2c572e22d                                                                                                                                         
                        at libstd/panicking.rs:310                                                                                                                                                                 
  12:     0x55f751c08469 - __rust_maybe_catch_panic                                                                                                                                                                
                        at libpanic_unwind/lib.rs:105                                                                                                                                                              
  13:     0x55f751bf19a5 - std::panicking::try::h3777f1f99a655493                                                                                                                                                  
                        at libstd/panicking.rs:289                                                                                                                                                                 
                         - std::panic::catch_unwind::hab6d1467193fc938                                                                                                                                             
                        at libstd/panic.rs:374                                                                                                                                                                     
                         - std::rt::lang_start_internal::h106bcdfb2e3beaf8                                                                                                                                         
                        at libstd/rt.rs:58                                                                                                                                                                         
  14:     0x55f75173a553 - main                                                                                                                                                                                    
  15:     0x7fb5c9cbd2b0 - __libc_start_main                                                                                                                                                                       
  16:     0x55f751732619 - _start                                                                                                                                                                                  
  17:                0x0 - <unknown>  
```

**Comment by greenaddress** [comment]
I think electrs maybe should fail with a user error message at start if it can't connect to the RPC and issue a call (i.e. getblockcount or similar). Once it is past that, maybe it should reconnect automatically with logs about it, possibly using some exponential retry capped at say 10 minutes and starting at 1 second or something like that.

**Comment by greenaddress** [comment]
More information based on some initial investigation:

it looks like the cookie changes at each restart, so rpc connection should re-read the cookie ideally.

**Comment by romanz** [comment]
@greenaddress May I help with this issue?

**Comment by greenaddress** [comment]
@romanz sorry I was away for a while and have a backlog of things.  I still want to contribute but I'm not very seasoned with Rust (yet!) so I'm also slow. I don't have anything meaning full yet for this issue so if you want to take over I'll be happy to help review or test.

**Comment by romanz** [comment]
No problem :)
I'll push some refactoring work to the `disconnect` branch - it should help us to handle disconnections.

**Comment by greenaddress** [comment]
I briefly tested https://github.com/romanz/electrs/commit/6efab9c8a20808c2a8f275bb1075ca299c3ff78d

My immediate test was trying to restart bitcoin core and that made it crash. I didn't try some disconnect (but i think i could simulate with a tunnel/port forwarding)

**Comment by romanz** [comment]
Thanks for testing!
Could you please attach the panic message?

**Comment by romanz** [comment]
@greenaddress Please try https://github.com/romanz/electrs/commit/9ee92c823f5e1d5f5c64b3fbe858566b2fde1d83 - AFAIK it should handle bitcoind restarts :)

**Comment by greenaddress** [comment]
@romanz yep fixed now in 91e4dd8b6fbee8252856127639d6885199e6e25d

**Comment by romanz** [comment]
Cool, thanks for testing!
